src/components/Datepicker/Datepicker.js

```diff
@@ -54,30 +74,97 @@ export class Datepicker extends React.Component {
   constructor(props, context) {
     super(props, context);
     this.prefix = (classes, passThrough) => prefixClasses(this.context.cssPrefix, classes, passThrough);
+    const { preselectedDate } = this.props;
 
     this.state = {
-      selectedDate: this.props.selectedDate ? moment(this.props.selectedDate) : null,
+      isOpen: false,
+      preselectedDate: preselectedDate ? moment(preselectedDate) : null,
+      inputDate: preselectedDate ? moment(preselectedDate).format(salesforceDateFormat).toString() : '',
       viewedDate: moment(),
     };
+  }
 
-    this.viewYear = (e) => {
-      this.setState({ viewedDate: this.state.viewedDate.set('year', e.target.value) });
-    };
-    this.viewMonth = (relativeMonth) => {
-      this.setState({ viewedDate: this.state.viewedDate.add(relativeMonth, 'month') });
-    };
-    this.selectDay = (day) => {
-      this.setState({ selectedDate: day });
-      this.props.onSelectDate(day.format('YYYY-MM-DD'));
-    };
-    this.resetViewToToday = () => this.setState({ viewedDate: moment() });
+  /**
+   * Callback when the year changes
+   * @param  {Object} e <select /> onChange event
+   * @return {void}     interacts with component state
+   */
+  onChangeYear = (e) => {
+    const { viewedDate } = this.state;
+
+    this.setState({ viewedDate: viewedDate.set('year', e.target.value) });
+  };
+
+  /**
+   * Callback when the month changes
+   * @param  {Number} relativeMonth one month forward or backward
+   * @return {void}                 interacts with component state
+   */
+  onChangeMonth = (relativeMonth) => {
+    const { viewedDate } = this.state;
+
+    this.setState({ viewedDate: viewedDate.add(relativeMonth, 'month') });
+  };
+
+  /**
+   * Callback when the month changes
+   * @param  {Object} day day as moment object
+   * @return {void}       interacts with component state and callback from props
+   */
+  onChangeDay = (day) => {
+    const { onToggle, onChangeValidDate } = this.props;
+    this.setState({
+      preselectedDate: day,
+      inputDate: moment(day).format(salesforceDateFormat).toString(),
+      isInputInvalid: false,
+    });
+
+    onToggle();
+    onChangeValidDate(day.format(iso8601DateFormat));
+  };
+
+  /**
+   * Callback when today reset link is cliecked
+   * @return {void} interacts with component state
+   */
+  onChangeToToday = () => this.setState({ viewedDate: moment() });
+
+  /**
+   * Callback when the input field changes
+   * @param  {String} value input value as string
+   * @return {void}         interacts with component state and callback from props
+   */
+  onChangeInputField = (value) => {
```

Comment:
In general: I recommend using a Subject-first naming pattern for functions, e.g. `onInputFieldChange`. That way you can easily group functions by subject, e.g. `onMyStuffChange` and `onMyStuffDelete` etc. Especially for prop types and component mounting that makes the code easier to read.